core/server/common/src/main/java/alluxio/master/journal/AsyncJournalWriter.java

```diff
@@ -99,49 +146,161 @@ public long appendEntry(JournalEntry entry) {
+   * Closes the async writer.
+   * PS: It initiates but does not wait for outstanding entries to be flushed.
    */
-  @SuppressWarnings("Duplicates")
-  public void flush(final long targetCounter) throws IOException, JournalClosedException {
-    if (targetCounter <= mFlushCounter.get()) {
-      return;
+  public void close() {
+    if (mFlushThread != null) {
+      mFlushThread.interrupt();
     }
-    // Using reentrant lock, since it seems to result in higher throughput than using 'synchronized'
-    try (LockResource lr = new LockResource(mFlushLock)) {
-      long startTime = System.nanoTime();
-      long flushCounter = mFlushCounter.get();
-      if (targetCounter <= flushCounter) {
-        // The specified counter is already flushed, so just return.
-        return;
+  }
+
+  /**
+   * A dedicated thread that goes over outstanding queue items and writes/flushes them. Other
+   * threads can track progress by submitting tickets via ::flush() call.
+   */
+  private void flushProc() {
+    /**
+     * Runs the loop until thread is interrupted.
+     */
+    boolean exit = false;
+    while (!exit) {
+
+      /**
+       * Wait to be woken up by flush requests as long as the queue is empty. Bail and drain the
+       * queue, if interrupted.
+       */
+      synchronized (mFlushThread) {
+        while (mQueue.isEmpty()) {
+          try {
+            mFlushThread.wait();
+          } catch (InterruptedException ie) {
+            // Time to exit.
+            exit = true;
+            break;
+          }
+        }
       }
-      long writeCounter = mWriteCounter.get();
-      while (targetCounter > writeCounter) {
-        for (;;) {
+
+      try {
+        long writeCounter = mWriteCounter.get();
+        long startTime = System.nanoTime();
+        boolean needFlush = false;
+
+        /**
+         * Write pending entries to journal.
+         */
+        while (!mQueue.isEmpty()) {
           // Get, but do not remove, the head entry.
           JournalEntry entry = mQueue.peek();
           if (entry == null) {
-            // No more entries in the queue. Break out of the infinite for-loop.
+            // No more entries in the queue. Break write session.
             break;
           }
           mJournalWriter.write(entry);
+          needFlush = true;
           // Remove the head entry, after the entry was successfully written.
           mQueue.poll();
           writeCounter = mWriteCounter.incrementAndGet();
 
-          if (writeCounter >= targetCounter) {
-            if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
-              // This thread has been writing to the journal for enough time. Break out of the
-              // infinite for-loop.
-              break;
+          if ((System.nanoTime() - startTime) >= mFlushBatchTimeNs) {
+            // This thread has been writing to the journal for enough time. Break out of the
+            // infinite while-loop.
+            break;
```

Comment:
Is it safe to do this before doing the following operations?

Reply:
It breaks from the write-loop. It'll still do flush and process the tickets at the end.
Is that what you mean?

Reply:
Oh got it, thanks.